src/Ledger/TokenAlgebra/ValueRelation.lagda

```diff
@@ -0,0 +1,464 @@
+
+
+
+    private
+     I : AssetID ⇀ Quantity
+     I = ι , (λ 0b 0b' → trans 0b (sym 0b'))
+
+    mh : IsMonoidMorphism (monoid Vcm) (monoid +-0-commutativeMonoid) _hasHowMuchAda
+    IsSemigroupMorphism.⟦⟧-cong (IsMonoidMorphism.sm-homo mh) {R} {R'} R⇔R' = goal
+     where
+     q q' : Quantity
+     q = R hasHowMuchAda
+     q' = R' hasHowMuchAda
+
+     RR' : (proj₁ R) ⇐⇒ (proj₁ R')
+     RR' = R⇔R'
+
+     Rq : (proj₁ R) (AdaPolicy , AdaName) q
+     Rq = proj₂ (AdaForAll R)
+
+     R'q' : (proj₁ R') (AdaPolicy , AdaName) q'
+     R'q' = proj₂ (AdaForAll R')
+
+     Rq' : (proj₁ R) (AdaPolicy , AdaName) q'
+     Rq' = proj₂ RR' R'q'
+
+     goal : R hasHowMuchAda ≡ R' hasHowMuchAda
+     goal = proj₂ R Rq Rq'
+
+    IsSemigroupMorphism.∙-homo (IsMonoidMorphism.sm-homo mh) Ru Rv = goal
+     where
+     q qu qv : Quantity
+     q = (Ru ⊕ Rv) hasHowMuchAda
+     ξ : (proj₁ (Ru ⊕ Rv)) (AdaPolicy , AdaName) q
+     ξ = proj₂ (AdaForAll (Ru ⊕ Rv))
+     qu = Ru hasHowMuchAda
+     qv = Rv hasHowMuchAda
+
+     α : proj₁ Ru (AdaPolicy , AdaName) (proj₁ ξ)
+     α = proj₁ (proj₂ (proj₂ (proj₂ (AdaForAll (Ru ⊕ Rv)))))
+     β : proj₁ Rv (AdaPolicy , AdaName) (proj₁ (proj₂ ξ))
+     β = proj₁ (proj₂ (proj₂ (proj₂ (proj₂ (AdaForAll (Ru ⊕ Rv))))))
+
+
+     quξ : qu ≡ proj₁ ξ
+     quξ = proj₂ Ru (proj₂ (AdaForAll Ru)) α
+     qvξ : qv ≡ proj₁ (proj₂ ξ)
+     qvξ = proj₂ Rv (proj₂ (AdaForAll Rv)) β
+
+     γ : proj₁ ξ + proj₁ (proj₂ ξ) ≡ qu + qv
+     γ = begin
+          proj₁ ξ + proj₁ (proj₂ ξ)  ≡⟨ cong (λ x → proj₁ ξ + x) (sym qvξ) ⟩
+          proj₁ ξ + qv               ≡⟨ cong (λ x → x + qv ) (sym quξ) ⟩
+          qu + qv                    ∎
+
+     ξ' : q ≡ proj₁ ξ + proj₁ (proj₂ ξ)
+     ξ' = proj₂ (proj₂ (proj₂ (proj₂ ξ)))
+
+     goal : q ≡ qu + qv
+     goal = trans ξ' γ
+
+    IsMonoidMorphism.ε-homo mh with I hasHowMuchAda ≟ 0
+    ... | yes p = p
+    ... | no _ = ι-left-unique{AdaPolicy , AdaName} (proj₂ (AdaForAll I)) _≡_.refl
+
+\end{code}
+
+`Val` is a finitely supported map. If `aid` is an `AssetID` and `v : Val`, then the quantity
+of assets contained in `v` with `AssetID` `aid` is `v aid`. Elements of `Val` are also
+referred to as asset bundles.
+
+Remarks.
+
+In the fls, `Val` is assumed to be a finitely supported map.  There are a number of ways we
+could manifest this assumption.
+1. Take the domain type to be finite and let `Val` be a total function on that domain.
+2. Let `Val` be a partial function `AssetID ⇀ Quantity` and either let
+   - `Quantity = ℤ ⊎ {⊥}` and `Val aid = ⊥` for all but finitely many `aid : AssetID` OR
+   - `Quantity = Maybe ℤ` and `Val aid = nothing` for all but finitely many `aid : AssetID`
+   (The two options are equivalent.)
+3. Use the fact that the set theory, as Andre has defined it, is inherently finite.
+
+In my view the first option, with a finite domain, seems the easiest to express and work with.
+
+Here are two alternative approaches that might be better:
+
+---------------------
```

Comment:
Anything below here is very implementation-y. If you want to keep this, you could make an appendix or something for this.

Reply:
I'll move that stuff to a separate module in a file under `TokenAlgebra`, called `DataStructures` or something, and leave it out of this PR, maybe working on it at some point in the future if/when we want to consider alternative implementations, optimizations, and the like.